### PR TITLE
Better implementation of `_shared_utils.results_path()` - using PyTest `request`

### PR DESCRIPTION
**Pull request by trexfeathers**
The previous implementation was flaky given we could not know where in the call stack the test actually was, and I have no idea how I could have got it working with PyTest parameters - probably something even more flaky.

This version uses publicly available attributes from the PyTest [`FixtureRequest`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.FixtureRequest), which seems much safer.

### CODE REVIEW

File: lib/iris/tests/_shared_utils.py [hunk 1]
```diff
@@ -207,17 +206,36 @@ def get_result_path(relative_path):
...
+        )
+        raise ValueError(message)
+
+
+def result_path(request: pytest.FixtureRequest, basename=None, ext=""):
     """Generate the path to a test result; from the calling file, class, method.
 
     Parameters
     ----------
+    request : pytest.FixtureRequest
```
Comment (ESadek-MO) [inline lib/iris/tests/_shared_utils.py:228]: Perhaps make it clear that all it needs to have is the word request, and pytest does the rest?

**Review by ESadek-MO — CHANGES_REQUESTED** [review]
LGTM, thanks @trexfeathers! 

Just a small suggestion, but otherwise awesome!

File: lib/iris/tests/_shared_utils.py (hunk 1 above)
Comment (trexfeathers) [reply lib/iris/tests/_shared_utils.py:228]: 6829c759d9017499dca3fd2bc185a2aaa83a8e02

**Review by ESadek-MO — APPROVED** [review]
Awesome, thanks!

**Pull request merged**